app/models/katello/concerns/smart_proxy_extensions.rb

```diff
@@ -443,7 +443,11 @@ def smart_proxy_service
       end
 
       def rhsm_url
-        if pulp_primary?
+        # Since Foreman 3.1 this setting is set
+        if (url = setting(SmartProxy::PULP3_FEATURE, 'rhsm_url'))
```

Comment:
Looks like this can resolve to nil and cause the tests to fail. While it might be good for some updated fixtures to reflect this, I think primarily this should not be allowed to resolve to nil and break.

Reply:
I think what is happening is you are shadowing the instance attribute with a local variable of the same name. The smart proxy has a `url` property, which this function is using. This if statement sets `url` to nil and then when this cascades down into the elsif blocks the value of url is nil and the `URI.parse` call fails:

```
irb(main):004:0> url = 'test';if (url = nil); puts 'hi'; else; puts url; end

=> nil
```

Reply:
Oh, that's a good point. I think this should be better then.